File: RepoPromptTests/CodeMap/Fixtures/go/smoke.go
// Smoke test fixture for Go codemap extraction
package user

import (
	"context"
	"errors"
	"sync"
	"time"
)

// UserRole represents the role of a user
type UserRole string

const (
	RoleAdmin  UserRole = "admin"
	RoleEditor UserRole = "editor"
	RoleViewer UserRole = "viewer"
	RoleGuest  UserRole = "guest"
)

// MaxUsers is the maximum number of users allowed
const MaxUsers = 1000

// DefaultTimeout for operations
var DefaultTimeout = 30 * time.Second

// User represents a user in the system
type User struct {
	ID        string    `json:"id"`
	Name      string    `json:"name"`
	Email     string    `json:"email"`
	Role      UserRole  `json:"role"`
	CreatedAt time.Time `json:"created_at"`
}

// DisplayName returns the formatted display name
func (u *User) DisplayName() string {
	return u.Name + " (" + string(u.Role) + ")"
}

// Validate checks if the user data is valid
func (u *User) Validate() error {
	if u.ID == "" {
		return errors.New("user ID cannot be empty")
	}
	if u.Name == "" {
		return errors.New("user name cannot be empty")
	}
	return nil
}

// DataProvider defines the interface for data access
type DataProvider[T any] interface {
	Fetch(ctx context.Context) ([]T, error)
	Save(ctx context.Context, item T) error
	Delete(ctx context.Context, id string) error
}

// UserService manages user operations
type UserService struct {
	mu     sync.RWMutex
	users  map[string]*User
	config Config
}

// Config holds service configuration
type Config struct {
	MaxUsers int
	Timeout  time.Duration
}

// NewUserService creates a new UserService instance
func NewUserService(cfg Config) *UserService {
	return &UserService{
		users:  make(map[string]*User),
		config: cfg,
	}
}

// Fetch retrieves all users
func (s *UserService) Fetch(ctx context.Context) ([]*User, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	users := make([]*User, 0, len(s.users))
	for _, u := range s.users {
		users = append(users, u)
	}
	return users, nil
}

// Save stores a user
func (s *UserService) Save(ctx context.Context, user *User) error {
	if err := user.Validate(); err != nil {
		return err
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	if len(s.users) >= s.config.MaxUsers {
		return errors.New("max users reached")
	}

	s.users[user.ID] = user
	return nil
}

// Delete removes a user by ID
func (s *UserService) Delete(ctx context.Context, id string) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	if _, exists := s.users[id]; !exists {
		return errors.New("user not found")
	}

	delete(s.users, id)
	return nil
}

// GetByRole returns users with the specified role
func (s *UserService) GetByRole(role UserRole) []*User {
	s.mu.RLock()
	defer s.mu.RUnlock()

	var result []*User
	for _, u := range s.users {
		if u.Role == role {
			result = append(result, u)
		}
	}
	return result
}

// CreateUser is a factory function for creating users
func CreateUser(name, email string, role UserRole) *User {
	return &User{
		ID:        generateID(),
		Name:      name,
		Email:     email,
		Role:      role,
		CreatedAt: time.Now(),
	}
}

// generateID creates a unique identifier
func generateID() string {
	return "user_" + time.Now().Format("20060102150405")
}

File: RepoPromptTests/CodeMap/Fixtures/rs/smoke.rs
// Smoke test fixture for Rust codemap extraction
use std::collections::HashMap;
use std::sync::{Arc, RwLock};
use std::error::Error;

/// User role enumeration
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum UserRole {
    Admin,
    Editor,
    Viewer,
    Guest,
}

impl Default for UserRole {
    fn default() -> Self {
        UserRole::Guest
    }
}

/// User data structure
#[derive(Debug, Clone)]
pub struct User {
    pub id: String,
    pub name: String,
    pub email: String,
    pub role: UserRole,
}

impl User {
    /// Creates a new user
    pub fn new(id: String, name: String, email: String) -> Self {
        Self {
            id,
            name,
            email,
            role: UserRole::default(),
        }
    }

    /// Returns the display name
    pub fn display_name(&self) -> String {
        format!("{} ({:?})", self.name, self.role)
    }

    /// Validates the user data
    pub fn validate(&self) -> Result<(), &'static str> {
        if self.id.is_empty() {
            return Err("User ID cannot be empty");
        }
        if self.name.is_empty() {
            return Err("User name cannot be empty");
        }
        Ok(())
    }
}

/// Trait for data providers
pub trait DataProvider<T> {
    fn fetch(&self) -> Result<Vec<T>, Box<dyn Error>>;
    fn save(&mut self, item: T) -> Result<(), Box<dyn Error>>;
    fn delete(&mut self, id: &str) -> Result<bool, Box<dyn Error>>;
}

/// Service configuration
#[derive(Debug, Clone)]
pub struct Config {
    pub max_users: usize,
    pub timeout_secs: u64,
}

impl Default for Config {
    fn default() -> Self {
        Self {
            max_users: MAX_USERS,
            timeout_secs: 30,
        }
    }
}

/// User service for managing users
pub struct UserService {
    users: Arc<RwLock<HashMap<String, User>>>,
    config: Config,
}

impl UserService {
    /// Creates a new UserService
    pub fn new(config: Config) -> Self {
        Self {
            users: Arc::new(RwLock::new(HashMap::new())),
            config,
        }
    }

    /// Gets users by role
    pub fn get_by_role(&self, role: UserRole) -> Vec<User> {
        let users = self.users.read().unwrap();
        users
            .values()
            .filter(|u| u.role == role)
            .cloned()
            .collect()
    }

    /// Returns the number of users
    pub fn count(&self) -> usize {
        self.users.read().unwrap().len()
    }
}

impl DataProvider<User> for UserService {
    fn fetch(&self) -> Result<Vec<User>, Box<dyn Error>> {
        let users = self.users.read().unwrap();
        Ok(users.values().cloned().collect())
    }

    fn save(&mut self, user: User) -> Result<(), Box<dyn Error>> {
        user.validate()?;
        
        let mut users = self.users.write().unwrap();
        if users.len() >= self.config.max_users {
            return Err("Max users reached".into());
        }
        
        users.insert(user.id.clone(), user);
        Ok(())
    }

    fn delete(&mut self, id: &str) -> Result<bool, Box<dyn Error>> {
        let mut users = self.users.write().unwrap();
        Ok(users.remove(id).is_some())
    }
}

/// Maximum number of users
pub const MAX_USERS: usize = 1000;

/// Default timeout in seconds
pub const DEFAULT_TIMEOUT: u64 = 30;

/// Creates a new user with generated ID
pub fn create_user(name: String, email: String, role: UserRole) -> User {
    User {
        id: generate_id(),
        name,
        email,
        role,
    }
}

/// Generates a unique ID
fn generate_id() -> String {
    format!("user_{}", std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_millis())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_user_creation() {
        let user = User::new("1".into(), "Test".into(), "test@example.com".into());
        assert_eq!(user.role, UserRole::Guest);
    }
}
